src/samples/audio/smart_amp_test_ipc4.c

```diff
@@ -447,178 +290,171 @@ static smart_amp_proc get_smart_amp_process(struct comp_dev *dev,
 	}
 }
 
-static int smart_amp_copy(struct comp_dev *dev)
+/* Only reset the src_input and fb_input if the num_input_buffers
+ * (get from module adapter)has changed.
+ */
+static void smart_amp_input_setup(struct processing_module *mod,
+				  struct input_stream_buffer *inputs,
+				  int32_t num_input_buffers)
 {
-	struct smart_amp_data *sad = comp_get_drvdata(dev);
-	struct comp_buffer __sparse_cache *source_buf = buffer_acquire(sad->source_buf);
-	struct comp_buffer __sparse_cache *sink_buf = buffer_acquire(sad->sink_buf);
-	uint32_t avail_passthrough_frames;
-	uint32_t avail_feedback_frames;
-	uint32_t avail_frames = 0;
-	uint32_t source_bytes;
-	uint32_t sink_bytes;
-	uint32_t feedback_bytes;
+	struct smart_amp_data *sad = module_get_private_data(mod);
+	struct comp_buffer *buf;
+	struct comp_buffer __sparse_cache *buffer_c;
+	int i;
 
-	comp_dbg(dev, "smart_amp_copy()");
+	if (sad->input_stream_num == num_input_buffers)
+		return;
 
-	/* available bytes and samples calculation */
-	avail_passthrough_frames =
-		audio_stream_avail_frames(&source_buf->stream,
-					  &sink_buf->stream);
+	sad->input_stream_num = num_input_buffers;
 
-	k_mutex_lock(&sad->lock, K_FOREVER);
-	if (sad->feedback_buf) {
-		struct comp_buffer __sparse_cache *buf = buffer_acquire(sad->feedback_buf);
+	if (num_input_buffers == 1) {
+		sad->src_input = &inputs[0];
+		sad->src_buf = attr_container_of(inputs[0].data,
+						 struct comp_buffer __sparse_cache,
+						 stream, __sparse_cache);
```

Comment:
no, we shouldn't even store acquired buffer pointers in any persistent objects. You shouldn't keep and reuse pointers to acquired buffers between calls to `.process()`. `.process()` is called every time with pointers to buffers, they should be used, not something stored from previous calls. This really should be redesigned.

Reply:
I see no need to store these buffer pointers at all anymore if we're going to look them up during every process invocation

Reply:
once I looked up this buffers every process, but @RanderWang suggested me to have optimization to check only when input stream numbers changed. And this can improve the performance of the hot code.

Reply:
I think its best to just look it up every time @andrula-song. The acquired pointer will be released by the module adapter after the process callback is done. It does not makes sense to hold to the pointer.

Reply:
Yes your are right, save these pointers here is not right. But look up every time really costs many cycles and make the performance worse, maybe I can find a better way to  avoid this check later.

Reply:
you can store unacquired pointers and re-acquire them. Acquiring them is cheap when there's no sharing - it's just an uncached-to-cached conversion, and that is basically just an OR or an AND and an OR